### DISCUSSION TITLE
What to cache between CI builds

### DISCUSSION THREAD

**Discussion by MarkCallow** [microsoft/vcpkg#20382] [Q&A] [answered]
I need to use `vcpkg` in my macOS CI build and I have to figure out what I should cache between builds as some of my app's dependencies take a _very_ long time to build. If I follow the default `vcpkg` installation instructions I end up with a clone of the vcpkg repo and all build files plus the `installed` directory are in subdirectories of that.

A simple answer is to cache the whole repo but that seems excessive. In my appveyor Windows builds I only need to cache `installed` and `%LocalAppData%\vcpkg`. What is the equivalent to the latter on macOS? Unlike the Windows builds, vcpkg is not installed in the CI environment and building it takes a long time so I want to cache that too and whatever is necessary to determine if it needs to be rebuilt.

Issues I have to deal with are that I can only cache directories, so no caching of just the `vcpkg` executable, and, if I cache the entire repo I have to do a `git clone` on the first build and a `cd <clone>; git pull` on subsequent builds.

**Comment by autoantwort** [comment, marked as answer]
The corresponding location is `~/.cache/vcpkg`. I think this is enough.

**Reply by MarkCallow** [reply]
What about the vcpkg executable. Is that cached in `~/.cache/vcpkg`?

Assuming not, I'm thinking of putting `buildtrees`, `installed` and `packages` outside the repo clone and then caching the repo clone. Will that work? Are there environment variables that can be used to specify the location for each of these? In the `bootstrap.sh` script I see use of VCPKG_DOWNLOADS for the first. There is no mention of the other 2 directories. I assume they and any related environment variables must be used in the vcpkg executable itself but I can't find the source for that. Where is it?

`strings vcpkg` reveals a `_VCPKG_INSTALLED_DIR` and `CURRENT_BUILDTREES_DIR` environment variables. The `_` on the front of the first makes me think it is not intended that vcpkg users use it. The lack of a non-generic prefix on the second makes me think likewise.

An alternative is to have the vcpkg executable built somewhere outside the repo clone. Looking at the bootstrap scripts they don't have an option to specify the vcpkg root so I don't see how to do that.

**Reply by autoantwort** [reply]
> What about the vcpkg executable. Is that cached in ~/.cache/vcpkg?

No, but I don't know if you really want to cache it, because it can change if you update vcpkg. At our project we simply committed the vcpkg executable (but this is windows only).

> Will that work?

You can do that if cloning is slow. 

> Are there environment variables that can be used to specify the location for each of these?

Only parameters (see `vcpkg help`)

> don't have an option to specify the vcpkg root so I don't see how to do that.

You can set the env var `VCPKG_ROOT`

**Reply by MarkCallow** [reply]
> The corresponding location is ~/.cache/vcpkg`. I think this is enough.

Do you mean there is no need to cache `installed`?

>  > What about the vcpkg executable. Is that cached in ~/.cache/vcpkg?
>
> No, but I don't know if you really want to cache it, because it can change if you update vcpkg. At our project we simply committed the vcpkg executable (but this is windows only).

I was hoping that if I cache the directory where it is built and in a future build clone the current repo and do a build that CMake would figure out if the cached executable actually needed rebuilding or not.

>>    don't have an option to specify the vcpkg root so I don't see how to do that.
>
> You can set the env var VCPKG_ROOT

`scripts/bootstrap.sh` doesn't pay any attention to VCPKG_ROOT. It builds the vcpkg executable and creates the vkpkg root with the executable and everything else inside it. The root is set to the directory in the ancestor chain of the `scripts` directory that contains the file .vkpkg-root.  That file is in the repo root.

**Reply by autoantwort** [reply]
> Do you mean there is no need to cache installed?

I don't think so. If you already have the binary cache you don't have to cache the installed folder. 

> I was hoping that if I cache the directory where it is built and in a future build clone the current repo and do a build that CMake would figure out if the cached executable actually needed rebuilding or not.

If I run `./bootstrap-vcpkg.sh` twice it does a full rebuild each time.

I don't understand your vcpkg root question. What problem do you want to solve?

**Reply by MarkCallow** [reply]
> If I run ./bootstrap-vcpkg.sh twice it does a full rebuild each time.

That's helpful to know. Thanks.

> I don't understand your vcpkg root question. What problem do you want to solve?

I am looking for a way to get the `vcpkg` executable built to a directory outside the git repo clone so I can cache it without also caching the repo clone. But given your tip about ./bootstrap-vcpkg.sh it seems caching won't save anything, if I want to keep the vcpkg executable up to date.

**Reply by MarkCallow** [reply]
Yes, there is no need to cache `installed`.

I'm building `vcpkg` every time in CI environments where it is not already installed. It takes time but given that CI only lets you cache entire directories I don't see any alternative.
